File: Leetcode/654_Maximum_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
        if(nums.size() == 0)
        {
            return nullptr;
        }
        
        TreeNode* root = construct(0, nums.size() -1, nums);
        
        return root;
    }
    
    TreeNode* construct(int start, int end, vector<int>& nums)
    {
        if(start > end)
        {
            return nullptr;
        }
        
        int ptr = start;
        int maxNum = nums[start];
        for(unsigned int i = start + 1; i <= end; i++)
        {
            if(maxNum < nums[i])
            {
                maxNum = nums[i];
                ptr = i;
            }
        }
        
        TreeNode * node = new TreeNode(maxNum);
        
        node->left = construct(start, ptr -1, nums);
        node->right = construct(ptr + 1, end, nums);
            
        return node;
    }
};

File: Leetcode/654_Maximum_Binary_Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var constructMaximumBinaryTree = function(nums) {
  if (!nums.length) return null;

  const [maxNum, maxIndex] = helper(nums);
  const root = new TreeNode(maxNum);
  root.left = constructMaximumBinaryTree(nums.slice(0, maxIndex));
  root.right = constructMaximumBinaryTree(nums.slice(maxIndex + 1));
  return root;
};

var helper = function(nums) {
  let maxNum = nums[0];
  let maxIndex = 0;
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] > maxNum) {
      maxNum = nums[i];
      maxIndex = i;
    }
  }
  return [maxNum, maxIndex];
};
